libs/ui-toolkit/src/components/callout/callout.tsx

```diff
@@ -1,23 +1,62 @@
+import type { ReactNode } from 'react';
 import classNames from 'classnames';
 import { getIntentShadow, Intent } from '../../utils/intent';
 import type { IconName } from '../icon';
 import { Icon } from '../icon';
 
-export interface CalloutProps {
+interface CalloutRootProps {
   children?: React.ReactNode;
   title?: React.ReactElement | string;
   intent?: Intent;
-  iconName?: IconName;
   headingLevel?: 1 | 2 | 3 | 4 | 5 | 6;
 }
 
-export function Callout({
-  children,
-  title,
-  intent = Intent.Help,
-  iconName,
-  headingLevel,
-}: CalloutProps) {
+interface CalloutWithoutIcon extends CalloutRootProps {
+  iconName?: never;
+  icon?: never;
+}
+
+interface CalloutPropsWithIconName extends CalloutRootProps {
+  children?: React.ReactNode;
+  title?: React.ReactElement | string;
+  intent?: Intent;
+  headingLevel?: 1 | 2 | 3 | 4 | 5 | 6;
+  iconName: IconName;
+  icon?: never;
+}
+
+interface CalloutPropsWithIcon extends CalloutRootProps {
+  children?: React.ReactNode;
+  title?: React.ReactElement | string;
+  intent?: Intent;
+  headingLevel?: 1 | 2 | 3 | 4 | 5 | 6;
+  iconName?: never;
+  icon: ReactNode;
+}
+
+type CalloutProps =
```

Comment:
Chore (blocking): I like the self-documenting code approach, but as far as I can tell, the benefit of having these different interfaces is that you've defined a situation where you never have an `iconName` and `icon` prop simultaneously. This isn't necessary, as your code in `getIconElement` handles that scenario. Therefore, we only need a single interface:

`interface CalloutProps {
  children?: React.ReactNode;
  title?: React.ReactElement | string;
  intent?: Intent;
  headingLevel?: 1 | 2 | 3 | 4 | 5 | 6;
  iconName?: IconName;
  icon?: ReactNode;
}`

Reply:
I think it is important to realize that my implementation is not equivalent with your suggestion. 
As a consumer of the Callout component, I'd like to avoid any potential confusion. If I do what you suggest here, that would allow to pass in both icon and iconName as props, but we never render both at the same time.
If we allow passing in both icon and iconName to the component, we would be left relying on the internal implementation to handle which would be overriding the other. In the runtime I am still making an arbitrary choice, of course, but it would not let tsc pass.
just had a look and realized you don't run typecheck in the pipelines, so not sure how much value this provides, other then assuming the devs have tsc hooked up in their IDE correctly and they see error messages in there.

Reply:
Ok cool, if you make the change to take advantage of extending CalloutRootProps in each interface then this gets a tick from me